nemo_aligner/algorithms/supervised.py

```diff
@@ -236,11 +236,11 @@ def set_max_steps(self):
         if (max_steps := self.cfg.get("max_steps", -1)) >= 0:
             self.max_steps = min(self.max_steps, max_steps)
 
-    def state_dict(self):
+    def state_dict(self, is_train_end=False):
         return {
             "step": self.step,
             "consumed_samples": self.consumed_samples,
-            "epoch": self.epoch,
+            "epoch": self.epoch + int(is_train_end),
```

Comment:
Can you explain what this is fixing? (for instance it would seem fine to me to have `epoch` be zero when saving at the end of a 1-epoch training)
(also probably that if this is needed for the supervised trainer then we also need to change it in other trainers as well)

Reply:
Yeah, so at the moment there's a pretty big oversight with when we save down our checkpoints, which is actually part of a much bigger problem in how we structure our training loops in general across all of Aligner. In this specific example which you have flagged, the checkpoint is written with `is_train_end=True` inside the inner loop which iterates on steps, but before you return to the outer loop where self.epoch is incremented. This is bad because you end up saving down the final checkpoint before self.epoch has been correctly incremented to its final value, so the checkpoint is saved with the wrong epoch number in the filename (which is then read again when resume training). Because the epoch number is incorrect in the filename, it is incorrectly restored at resume-time, which then adds an extra epoch to training when it should not.

The problem is we shouldn't be saving down the final checkpoint inside the inner loop, it should be happening in the outer loop once self.epoch has been correctly incremented.

However, fixing that is more difficult because it would interfere with saving non-final checkpoints. Which brings me to my next point: the general structure of our training logic has some flaws, using this outer for-loop for epochs, and inner for loop for step per epoch. The reason this is bad is because it causes all manner of problems for correct PTL use of max_epochs and max_steps, making our codebase not PTL conformant. Ideally, we should rewrite the core training loop to be a single forever loop, and then it breaks when it reaches either max_steps or max_epochs, and it keeps track of which epoch it is on via a counter which is incremented by modulo to the dataloader length.

Reply:
I see, thanks for the detailed explanation! What about we replace `self.epoch` with a property that computes it dynamically from `self.step` and `self._train_dataloader_len`? (instead of trying to save / load it back)

I'm a bit uncomfortablel with the `+ int(is_train_end)` because (1) it makes the code less readable, and (2) it may actually be incorrect in some situations (for instance if we manually set `max_steps` in the config in such a way that we don't end training at the end of an epoch).

Reply:
I have thought of that, but `self._train_dataloader_len` isn't a constant, it changes based on the consumed samples/total samples as fed into the RandomBatchSampler. It starts at the full len(dataloader) then decays by consumed samples until it hits zero and then resets back for the next shuffled batch, ad infinitum. You can look at the RandomBatchSampler code to see what I mean. The `__len__` method there is dynamic based on where you are in your percentage of consumed samples vs total samples.

Reply:
The real fix here is to rewrite that training loop, it's just not correct as is because the self.epoch isn't being updated where it needs to be updated for the downstream logic to work correctly

Reply:
Wait, I have an idea for a fix, let's move self.save | is_train_end, outside of both loops, that should work fine.

Reply:
I have implemented my fix where I move the self.save call outside both for loops

Reply:
I assume that should work, but as discussed over Slack I think it'd be best if we avoided making the training loop more complex.

I have an alternative suggestion here: eae530c5 (relies on the `num_steps_per_epoch` from that commit I discussed in the other thread). I still need to test it properly but it seems the code would be cleaner that way, what do you think?

(NB: ideally I would have removed `epoch` from the state dict but PTL needs it in `monitor_candidates` and it seems better to me to keep it there along with the other trainer-specific metrics rather than manually add it in `save()`)

Reply:
what if we:

1. compute `self.max_steps` using the epoch number to get total max steps
2. edit [check_progress](https://github.com/NVIDIA/NeMo-Aligner/blob/degert/multi-epochs/nemo_aligner/utils/trainer_utils.py#L40C5-L40C19) to only return `is_train_end=True` when training is actually completed

we can remove the outer epoch loop and compute epoch using @odelalleau 's [change](https://github.com/NVIDIA/NeMo-Aligner/commit/eae530c59c4694fcc58140542fd1789ce0c964b7#diff-e38954660a998f69fbd8f2611817c6b191f3f62413dc010d67de63b02a755ccbR522)

Reply:
> what if we:
> 
> 1. compute `self.max_steps` using the epoch number to get total max steps
> 2. edit [check_progress](https://github.com/NVIDIA/NeMo-Aligner/blob/degert/multi-epochs/nemo_aligner/utils/trainer_utils.py#L40C5-L40C19) to only return `is_train_end=True` when training is actually completed

I'm not sure what you mean by that. `max_steps` should already be the total max steps (though current code is incorrect, but I proposed a fix in 70edf5acdfb7dd21b07265cad653e0ca4c202c8d). And `check_progress()` is only supposed to set `is_train_end` to True when training is actually completed (`is_train_end = step == max_steps`).

Reply:
i see, if this is the case then perhaps we should remove the [epoch loop](https://github.com/NVIDIA/NeMo-Aligner/commit/eae530c59c4694fcc58140542fd1789ce0c964b7#diff-e38954660a998f69fbd8f2611817c6b191f3f62413dc010d67de63b02a755ccbR370) all together since we already compute epoch in your change. Do you think that makes it more complicated?

Reply:
I have pushed Olivier's suggested fixes to DPO/PPO/Supervised